planning/behavior_velocity_planner/autoware_behavior_velocity_virtual_traffic_light_module/src/experimental/scene.cpp

```diff
@@ -14,88 +14,134 @@
 
 #include "scene.hpp"
 
-#include "../utils.hpp"
-
-#include <autoware/behavior_velocity_planner_common/utilization/arc_lane_util.hpp>
 #include <autoware/behavior_velocity_planner_common/utilization/util.hpp>
-#include <autoware/motion_utils/trajectory/trajectory.hpp>
+#include <autoware/lanelet2_utils/conversion.hpp>
+#include <autoware/trajectory/utils/crossed.hpp>
+#include <autoware/trajectory/utils/find_nearest.hpp>
 #include <autoware_utils/ros/marker_helper.hpp>
 
+#include <tier4_v2x_msgs/msg/key_value.hpp>
+
 #include <memory>
 #include <string>
 #include <vector>
 
 namespace autoware::behavior_velocity_planner::experimental
 {
-using virtual_traffic_light::calcCenter;
-using virtual_traffic_light::calcHeadPose;
-using virtual_traffic_light::createKeyValue;
-using virtual_traffic_light::findLastCollisionBeforeEndLine;
-using virtual_traffic_light::insertStopVelocityFromStart;
-using virtual_traffic_light::SegmentIndexWithOffset;
-using virtual_traffic_light::SegmentIndexWithPoint;
-using virtual_traffic_light::toAutowarePoints;
+namespace
+{
+tier4_v2x_msgs::msg::KeyValue createKeyValue(const std::string & key, const std::string & value)
+{
+  return tier4_v2x_msgs::build<tier4_v2x_msgs::msg::KeyValue>().key(key).value(value);
+}
+
+std::optional<double> findLastCollisionBeforeLine(
+  const Trajectory & path, const double end_line_s, const lanelet::ConstLineString3d & line)
+{
+  const auto collisions = [&]() {
+    auto cropped_path = path;
+    cropped_path.crop(0., std::min(end_line_s, cropped_path.length()));
+    return autoware::experimental::trajectory::crossed(cropped_path, line);
+  }();
+
+  if (collisions.empty()) {
+    return std::nullopt;
+  }
+
+  return collisions.back();
+}
+
+std::optional<double> findLastCollisionBeforeLine(
+  const Trajectory & path, const double end_line_s, const lanelet::ConstLineStrings3d & lines)
+{
+  for (const auto & line : lines) {
+    if (const auto collision = findLastCollisionBeforeLine(path, end_line_s, line)) {
```

Comment:
Just for clarification:

```suggestion
    if (const auto collision = findLastCollisionBeforeLine(path, end_line_s, line); collision) {
```